### ISSUE TITLE
GeoSPARQL -  geof:sfIntersects function issue when left-hand side geometry is rectangle shape

### ISSUE THREAD

**Issue by LorenzBuehmann** [apache/jena#2219]
### Version

4.10.0 and 5.0.0-SNAPSHOT

### What happened?

Calling the GeoSPARQL function `geof:sfIntersects` with e.g. `FILTER(geof:sfIntersects(?g1, ?g2))` can lead to an exception when the first geometry `g1`is an rectangle. See the following code to reproduce:

``` java
package org.apache.jena.geosparql;

import org.apache.jena.geosparql.implementation.GeometryWrapper;
import org.apache.jena.geosparql.implementation.vocabulary.GeoSPARQL_URI;
import org.apache.jena.graph.Node;
import org.apache.jena.riot.system.PrefixMap;
import org.apache.jena.riot.system.PrefixMapFactory;
import org.apache.jena.sparql.util.NodeFactoryExtra;

public class GeometryIntersectionBug {

    public static void main(String[] args) throws Exception {
        PrefixMap pm = PrefixMapFactory.create(GeoSPARQL_URI.getPrefixes());

        // a hexagon
        String polygonLitStr = "\"POLYGON ((10.0656923 55.3948787, 9.3745308 53.9755137, 11.2451705 53.0320397, 13.8395286 53.4872452, 14.6402066 54.9021819, 12.739823 55.8673267, 10.0656923 55.3948787))\"^^geo:wktLiteral";
        Node polygonNode = NodeFactoryExtra.parseNode(polygonLitStr, pm);
        GeometryWrapper polygonGeom = GeometryWrapper.extract(polygonNode);

        // a rectangle
        String rectangleLitStr = "\"POLYGON((6.2520337 47.7361897, 6.2520337 53.7746148, 14.0584217 53.7746148, 14.0584217 47.7361897, 6.2520337 47.7361897))\"^^geo:wktLiteral";
        Node rectangleNode = NodeFactoryExtra.parseNode(rectangleLitStr, pm);
        GeometryWrapper rectangleGeom = GeometryWrapper.extract(rectangleNode);

        // check if both geometries intersect
        boolean intersects1  = polygonGeom.intersects(rectangleGeom); // works

        boolean intersects2 = rectangleGeom.intersects(polygonGeom); // error

    }
}
```
It happens because the underlying JTS API does use an optimized algorithm for rectangles to terminate as early as possible. During this algorithm some new `Coordinate` objects will be created which make use of some Jena specific implementation which assumes not only XY geometries but also the Z axis. This results in an exception.

### Relevant output and stacktrace

```shell
Exception in thread "main" java.lang.IllegalArgumentException: CoordinateXY dimension 2 does not support z-ordinate
	at org.locationtech.jts.geom.CoordinateXY.setZ(CoordinateXY.java:99)
	at org.apache.jena.geosparql.implementation.jts.CustomCoordinateSequence.getCoordinate(CustomCoordinateSequence.java:397)
	at org.locationtech.jts.operation.predicate.RectangleIntersectsSegmentVisitor.checkIntersectionWithSegments(RectangleIntersects.java:327)
	at org.locationtech.jts.operation.predicate.RectangleIntersectsSegmentVisitor.checkIntersectionWithLineStrings(RectangleIntersects.java:315)
	at org.locationtech.jts.operation.predicate.RectangleIntersectsSegmentVisitor.visit(RectangleIntersects.java:308)
	at org.locationtech.jts.geom.util.ShortCircuitedGeometryVisitor.applyTo(ShortCircuitedGeometryVisitor.java:35)
	at org.locationtech.jts.operation.predicate.RectangleIntersects.intersects(RectangleIntersects.java:110)
	at org.locationtech.jts.operation.predicate.RectangleIntersects.intersects(RectangleIntersects.java:58)
	at org.locationtech.jts.geom.prep.PreparedPolygon.intersects(PreparedPolygon.java:84)
	at org.apache.jena.geosparql.implementation.GeometryWrapper.intersects(GeometryWrapper.java:882)
	at org.apache.jena.geosparql.GeometryIntersectionBug.main(GeometryIntersectionBug.java:29)
```

### Are you interested in making a pull request?

Maybe

**Comment by neumarcx** [comment]
thank you for reporting this Lorenz. As GeoSPARQL 1.0 is inherently 2D we should just enforce this property at the function level as well. It should not prevent users from capturing the z dimension of course.

I was able to reproduce this bug in the current 4.10.0 and 5.0.0-SNAPSHOT version in code and I can confirm it's a Jena / SIS combination issue and not JTS related.  FYI this bug is also not featuring in my last development branch (fuseki-geosparql 4.8.0 version).

**Comment by LorenzBuehmann** [comment]
Hi Marco.

Yep, this is indeed no JTS issue but the custom implementation of an optimized `CoordinateSequence` which handles Z axis and assumes that this is always possible.  
From what I can see in the JTS code why it happens for rectangles, because it's more efficient to evaluate whether an intersection exist in that case. It's interesting that it doesn't always fail because in the code there are two more check in before checking the line segments of the rectangle. We recognized the error only because we added H3 support to the Jena spatial layer via property functions, but needed a post filter on a bounding box which in fact is a rectangle.
It's also lucky that we passed the rectangle as first argument, otherwise a different algorithm will be used by JTS.

It's easy to fix in the code I think just adding an `if` condition check on the type of the `Coordinate` object (XYZ or not).
I'll provide a PR soon.

**Comment by neumarcx** [comment]
> Hi Marco.
> 
> Yep, this is indeed no JTS issue but the custom implementation of an optimized `CoordinateSequence` which handles Z axis and assumes that this is always possible. From what I can see in the JTS code why it happens for rectangles, because it's more efficient to evaluate whether an intersection exist in that case. It's interesting that it doesn't always fail because in the code there are two more check in before checking the line segments of the rectangle. We recognized the error only because we added H3 support to the Jena spatial layer via property functions, but needed a post filter on a bounding box which in fact is a rectangle. It's also lucky that we passed the rectangle as first argument, otherwise a different algorithm will be used by JTS.
> 
> It's easy to fix in the code I think just adding an `if` condition check on the type of the `Coordinate` object (XYZ or not). I'll provide a PR soon.

yes, very good Lorenz, I was able to patch my local copy in CustomCoordinateSequence and got it to work as well. I am still wondering about side effects. Also the unpatched version seems to work fine with the other topological relationships (overlaps, touches, disjoint etc) if isRectangle is true. The only call that fails is sfIntersects when the domain of the function qualifies as a rectangle.

**Comment by LorenzBuehmann** [comment]
I don't see any negative implications, for me this is just fixing a bug - indeed I might be wrong.

The crucial part in JTS is in class 

Here they do

``` java
private void checkIntersectionWithSegments(LineString testLine)
  {
    CoordinateSequence seq1 = testLine.getCoordinateSequence();
    Coordinate p0 = seq1.createCoordinate();
    Coordinate p1 = seq1.createCoordinate();
    for (int j = 1; j < seq1.size(); j++) {
      seq1.getCoordinate(j - 1, p0);
      seq1.getCoordinate(j,     p1);

      if (rectIntersector.intersects(p0, p1)) {
        hasIntersection = true;
        return;
      }
    }
  }
```
the creation of new `Coordinates` comes from the default method in the `CoordinateSequence` interface and does nothing more than taking into account the dimensions of coordinate and the measure:

``` java
default Coordinate createCoordinate() {
    return Coordinates.create(getDimension(), getMeasures());
  }
```

Both are implemented in `CustomCoordinateSequence` correctly, i.e. the dimension is 2 and measures is 0. 
I'm sure it was simply forgotten to do a sanity check in the `getCoordinate` method

``` java
@Override
    public void getCoordinate(int index, Coordinate coord) {
        coord.setX(x[index]);
        coord.setY(y[index]);
        coord.setZ(z[index]);
    }
``` 
and by the way for some reason `setM` was not called - maybe also forgotten.
In any case checking for either the class type or just the dimension should be the correct way in my opinion.
So either explicitly check for the Java type or for combination of dimension/measure should at least fix the error. In JTS they do it by

``` java
public static Coordinate create(int dimension, int measures)
  {
    if (dimension == 2) {
      return new CoordinateXY();
    } else if (dimension == 3 && measures == 0) {
      return new Coordinate();
    } else if (dimension == 3 && measures == 1) {
      return new CoordinateXYM();
    } else if (dimension == 4 && measures == 1) {
      return new CoordinateXYZM();
    }
    return new Coordinate();
  }
```
It's confusing how JTS does handle the case `dim=3` and `measure=1` as Z seems to be ignored: 
> This data object is suitable for use with coordinate sequences with dimension = 3 and measures = 1.
The Coordinate.z field is visible, but intended to be ignored


Long story short, we could adapt this with
``` java
@Override
    public void getCoordinate(int index, Coordinate coord) {
        coord.setX(x[index]);
        coord.setY(y[index]);

        if (coordinateDimension == 3 && measuresDimension == 0) {
            coord.setZ(z[index]);
        } else if (coordinateDimension == 3 && measuresDimension == 1) {
            coord.setM(m[index]);
        } else if (coordinateDimension == 4 && measuresDimension == 1) {
            coord.setZ(z[index]);
            coord.setM(m[index]);
        }
    }

```


Opinions?

**Comment by neumarcx** [comment]
This is well motivated and proposed. In my patch I have just ignored the z data in the function call.  But the cases are a better approach here. With side effects I was referring to the root cause analysis. I would think that it would make sense to add further test cases to the build as well.
